File: module9praktika.py
from abc import ABC, abstractmethod

class RoomBookingSystem:
    def book_room(self, guest_name):
        print(f"Бронирование номера для гостя: {guest_name}")

    def cancel_booking(self, guest_name):
        print(f"Отмена бронирования для гостя: {guest_name}")


class RestaurantSystem:
    def book_table(self, guest_name):
        print(f"Стол забронирован для гостя: {guest_name}")

    def order_food(self, guest_name, dishes):
        print(f"Заказ еды для {guest_name}: {', '.join(dishes)}")


class EventManagementSystem:
    def organize_event(self, event_name):
        print(f"Организовано мероприятие: {event_name}")

    def book_equipment(self, equipment_list):
        print(f"Забронировано оборудование: {', '.join(equipment_list)}")


class CleaningService:
    def schedule_cleaning(self, room_number):
        print(f"Уборка запланирована для комнаты №{room_number}")

    def clean_now(self, room_number):
        print(f"Уборка выполнена в комнате №{room_number}")


class TaxiService:
    def order_taxi(self, guest_name):
        print(f"Такси вызвано для гостя: {guest_name}")


# ФАСАД
class HotelFacade:
    def __init__(self):
        self.booking = RoomBookingSystem()
        self.restaurant = RestaurantSystem()
        self.events = EventManagementSystem()
        self.cleaning = CleaningService()
        self.taxi = TaxiService()

    def book_room_with_services(self, guest_name, dishes):
        print("\n Сценарий: Бронирование номера с едой и уборкой ")
        self.booking.book_room(guest_name)
        self.restaurant.order_food(guest_name, dishes)
        self.cleaning.schedule_cleaning(101)
        print("Все услуги успешно оформлены")

    def organize_event_with_rooms(self, event_name, participants, equipment):
        print("\n Сценарий: Организация мероприятия ")
        self.events.organize_event(event_name)
        self.events.book_equipment(equipment)
        for p in participants:
            self.booking.book_room(p)
        print("Мероприятие и бронирования участников оформлены")

    def book_table_with_taxi(self, guest_name):
        print("\n Сценарий: Бронирование стола с вызовом такси ")
        self.restaurant.book_table(guest_name)
        self.taxi.order_taxi(guest_name)
        print("Стол и такси оформлены")

    def cancel_booking(self, guest_name):
        print("\n Отмена бронирования ")
        self.booking.cancel_booking(guest_name)

    def clean_on_request(self, room_number):
        print("\n Уборка по запросу ")
        self.cleaning.clean_now(room_number)

class OrganizationComponent(ABC):
    @abstractmethod
    def display(self, indent=0):
        pass

    @abstractmethod
    def get_budget(self):
        pass

    @abstractmethod
    def get_employee_count(self):
        pass


class Employee(OrganizationComponent):
    def __init__(self, name, position, salary):
        self.name = name
        self.position = position
        self.salary = salary

    def display(self, indent=0):
        print(" " * indent + f"{self.position}: {self.name} (Зарплата: {self.salary})")

    def get_budget(self):
        return self.salary

    def get_employee_count(self):
        return 1

    def set_salary(self, new_salary):
        self.salary = new_salary
        print(f"Зарплата сотрудника {self.name} изменена на {new_salary}")


class Contractor(Employee):
    def get_budget(self):
        return 0  


class Department(OrganizationComponent):
    def __init__(self, name):
        self.name = name
        self.children = []

    def add(self, component):
        if component not in self.children:
            self.children.append(component)

    def remove(self, component):
        if component in self.children:
            self.children.remove(component)

    def display(self, indent=0):
        print(" " * indent + f"[Отдел] {self.name}")
        for child in self.children:
            child.display(indent + 4)

    def get_budget(self):
        return sum(child.get_budget() for child in self.children)

    def get_employee_count(self):
        return sum(child.get_employee_count() for child in self.children)

    def find_employee(self, name):
        for child in self.children:
            if isinstance(child, Employee) and child.name == name:
                return child
            elif isinstance(child, Department):
                found = child.find_employee(name)
                if found:
                    return found
        return None


if __name__ == "__main__":
    # --- ФАСАД ---
    hotel = HotelFacade()
    hotel.book_room_with_services("Иван Кличко", ["Пицца", "Кофе"])
    hotel.organize_event_with_rooms("IT Конференция", ["Анна", "Борян"], ["Проектор", "Микрофон"])
    hotel.book_table_with_taxi("Сергей")
    hotel.clean_on_request(203)
    hotel.cancel_booking("Иван Кличко")

    # --- КОМПОНОВЩИК ---
    print("\n================================")
    print("    КОРПОРАТИВНАЯ СТРУКТУРА")
    print("================================")

    e1 = Employee("Иван Кличко", "Менеджер", 120000)
    e2 = Employee("Анна Хилькевич", "Разработчик", 150000)
    e3 = Contractor("Дмитрий Нагиев", "Тестировщик (контрактор)", 70000)
    e4 = Employee("Мария Абрамович", "HR", 90000)

    dev_dept = Department("Отдел Разработки")
    hr_dept = Department("Отдел HR")
    main_office = Department("Главный офис")

    dev_dept.add(e2)
    dev_dept.add(e3)
    hr_dept.add(e4)
    main_office.add(e1)
    main_office.add(dev_dept)
    main_office.add(hr_dept)

    print("\n Структура организации ")
    main_office.display()
    print(f"\nОбщий бюджет: {main_office.get_budget()} руб.")
    print(f"Всего сотрудников: {main_office.get_employee_count()}")

    print("\n Поиск сотрудника ")
    found = main_office.find_employee("Анна Хилькевич")
    if found:
        found.display()
    else:
        print("Сотрудник не найден")

File: module9praktika.cs
using System;
using System.Collections.Generic;

public class RoomBookingSystem
{
    public void BookRoom(string guestName)
    {
        Console.WriteLine($"Бронирование номера для гостя: {guestName}");
    }

    public void CancelBooking(string guestName)
    {
        Console.WriteLine($"Отмена бронирования для гостя: {guestName}");
    }
}

public class RestaurantSystem
{
    public void BookTable(string guestName)
    {
        Console.WriteLine($"Стол забронирован для гостя: {guestName}");
    }

    public void OrderFood(string guestName, List<string> dishes)
    {
        Console.WriteLine($"Заказ еды для {guestName}: {string.Join(", ", dishes)}");
    }
}

public class EventManagementSystem
{
    public void OrganizeEvent(string eventName)
    {
        Console.WriteLine($"Организовано мероприятие: {eventName}");
    }

    public void BookEquipment(List<string> equipmentList)
    {
        Console.WriteLine($"Забронировано оборудование: {string.Join(", ", equipmentList)}");
    }
}

public class CleaningService
{
    public void ScheduleCleaning(int roomNumber)
    {
        Console.WriteLine($"Уборка запланирована для комнаты №{roomNumber}");
    }

    public void CleanNow(int roomNumber)
    {
        Console.WriteLine($"Уборка выполнена в комнате №{roomNumber}");
    }
}

public class TaxiService
{
    public void OrderTaxi(string guestName)
    {
        Console.WriteLine($"Такси вызвано для гостя: {guestName}");
    }
}

public class HotelFacade
{
    private RoomBookingSystem booking = new RoomBookingSystem();
    private RestaurantSystem restaurant = new RestaurantSystem();
    private EventManagementSystem events = new EventManagementSystem();
    private CleaningService cleaning = new CleaningService();
    private TaxiService taxi = new TaxiService();

    public void BookRoomWithServices(string guestName, List<string> dishes)
    {
        Console.WriteLine("\n Сценарий: Бронирование номера с едой и уборкой ");
        booking.BookRoom(guestName);
        restaurant.OrderFood(guestName, dishes);
        cleaning.ScheduleCleaning(101);
        Console.WriteLine("Все услуги успешно оформлены");
    }

    public void OrganizeEventWithRooms(string eventName, List<string> participants, List<string> equipment)
    {
        Console.WriteLine("\n Сценарий: Организация мероприятия ");
        events.OrganizeEvent(eventName);
        events.BookEquipment(equipment);
        foreach (string p in participants)
        {
            booking.BookRoom(p);
        }
        Console.WriteLine("Мероприятие и бронирования участников оформлены");
    }

    public void BookTableWithTaxi(string guestName)
    {
        Console.WriteLine("\n Сценарий: Бронирование стола с вызовом такси ");
        restaurant.BookTable(guestName);
        taxi.OrderTaxi(guestName);
        Console.WriteLine("Стол и такси оформлены");
    }

    public void CancelBooking(string guestName)
    {
        Console.WriteLine("\n Отмена бронирования ");
        booking.CancelBooking(guestName);
    }

    public void CleanOnRequest(int roomNumber)
    {
        Console.WriteLine("\n Уборка по запросу ");
        cleaning.CleanNow(roomNumber);
    }
}
public abstract class OrganizationComponent
{
    public abstract void Display(int indent = 0);
    public abstract double GetBudget();
    public abstract int GetEmployeeCount();
}

public class Employee : OrganizationComponent
{
    public string Name { get; set; }
    public string Position { get; set; }
    public double Salary { get; set; }

    public Employee(string name, string position, double salary)
    {
        Name = name;
        Position = position;
        Salary = salary;
    }

    public override void Display(int indent = 0)
    {
        Console.WriteLine(new string(' ', indent) + $"{Position}: {Name} (Зарплата: {Salary})");
    }

    public override double GetBudget() => Salary;

    public override int GetEmployeeCount() => 1;

    public void SetSalary(double newSalary)
    {
        Salary = newSalary;
        Console.WriteLine($"Зарплата сотрудника {Name} изменена на {newSalary}");
    }
}

public class Contractor : Employee
{
    public Contractor(string name, string position, double salary)
        : base(name, position, salary) { }

    public override double GetBudget() => 0;

public class Department : OrganizationComponent
{
    public string Name { get; set; }
    private List<OrganizationComponent> children = new List<OrganizationComponent>();

    public Department(string name)
    {
        Name = name;
    }

    public void Add(OrganizationComponent component)
    {
        if (!children.Contains(component))
            children.Add(component);
    }

    public void Remove(OrganizationComponent component)
    {
        if (children.Contains(component))
            children.Remove(component);
    }

    public override void Display(int indent = 0)
    {
        Console.WriteLine(new string(' ', indent) + $"[Отдел] {Name}");
        foreach (var child in children)
            child.Display(indent + 4);
    }

    public override double GetBudget()
    {
        double total = 0;
        foreach (var child in children)
            total += child.GetBudget();
        return total;
    }

    public override int GetEmployeeCount()
    {
        int count = 0;
        foreach (var child in children)
            count += child.GetEmployeeCount();
        return count;
    }

    public Employee FindEmployee(string name)
    {
        foreach (var child in children)
        {
            if (child is Employee emp && emp.Name == name)
                return emp;
            else if (child is Department dept)
            {
                var found = dept.FindEmployee(name);
                if (found != null)
                    return found;
            }
        }
        return null;
    }
}

public class Program
{
    public static void Main()
    {
        HotelFacade hotel = new HotelFacade();
        hotel.BookRoomWithServices("Иван Кличко", new List<string> { "Пицца", "Кофе" });
        hotel.OrganizeEventWithRooms("IT Конференция", new List<string> { "Анна", "Борян" }, new List<string> { "Проектор", "Микрофон" });
        hotel.BookTableWithTaxi("Сергей");
        hotel.CleanOnRequest(203);
        hotel.CancelBooking("Иван Кличко");

        Console.WriteLine("    КОРПОРАТИВНАЯ СТРУКТУРА");
     
        Employee e1 = new Employee("Иван Кличко", "Менеджер", 120000);
        Employee e2 = new Employee("Анна Хилькевич", "Разработчик", 150000);
        Contractor e3 = new Contractor("Дмитрий Нагиев", "Тестировщик (контрактор)", 70000);
        Employee e4 = new Employee("Мария Абрамович", "HR", 90000);

        Department devDept = new Department("Отдел Разработки");
        Department hrDept = new Department("Отдел HR");
        Department mainOffice = new Department("Главный офис");

        devDept.Add(e2);
        devDept.Add(e3);
        hrDept.Add(e4);
        mainOffice.Add(e1);
        mainOffice.Add(devDept);
        mainOffice.Add(hrDept);

        Console.WriteLine("\n Структура организации ");
        mainOffice.Display();
        Console.WriteLine($"\nОбщий бюджет: {mainOffice.GetBudget()} руб.");
        Console.WriteLine($"Всего сотрудников: {mainOffice.GetEmployeeCount()}");

        Console.WriteLine("\n Поиск сотрудника ");
        var found = mainOffice.FindEmployee("Анна Хилькевич");
        if (found != null)
            found.Display();
        else
            Console.WriteLine("Сотрудник не найден");
    }
}
